### ISSUE TITLE
comm error: bincode/io error: io error: frame size too big

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#3735]
The below SQL produces an error after a minute or two.

ERROR:  comm error: bincode/io error: io error: frame size too big

```
CREATE VIEW "materialize"."public"."view_1" ("col_1", "col_2", "col_3", "col_4", "col_5") AS SELECT ('dg8ZKzISR'::text <> 'Acx4O5zjN'::text) AS "ident_717", '09:27:08'::time::time AS "ident_718", 0.67960954::real AS "ident_719", 0.7592757308433459::double AS "ident_720", 0.491091196564053::double AS "ident_721" FROM "materialize"."public"."pseudo_source" AS "tab_561" GROUP BY "tab_561"."value";

CREATE VIEW "materialize"."public"."view_2" ("col_6", "col_7") AS SELECT 223952287012313072::bigint AS "ident_1299", '1744-08-20'::date::date AS "ident_1300" FROM "materialize"."public"."view_1" AS "tab_1106";

CREATE VIEW "materialize"."public"."view_4" ("col_9", "col_10") AS SELECT 'xq'::text AS "ident_1366", '2039-11-25 07:29:11+0000'::timestamp::timestamp AS "ident_1367" FROM "materialize"."public"."pseudo_source" AS "tab_1161", "materialize"."public"."pseudo_source" AS "tab_1166", "materialize"."public"."pseudo_source" AS "tab_1167", "materialize"."public"."view_2" AS "tab_1168", "materialize"."public"."pseudo_source" AS "tab_1169" WHERE true::boolean;

SELECT CAST('E3aul4JLB' AS TEXT) AS ident_1849, CAST(0.033044189132852675 AS DOUBLE) AS ident_1850 FROM view_4 AS tab_1506, view_4 AS tab_1507 OFFSET 23 ROWS;
```

**Comment by mjibson** [comment]
Running the 0.3.1 docker image.

**Comment by frankmcsherry** [comment]
Thanks! Do you have any information on what `"materialize"."public"."pseudo_source"` looks like, in particular its cardinality? The query ends up being a 10-way cross join of this relation with itself. It doesn't seem to look at any of the columns, so it shouldn't actually be a problematic query, but the cardinality of the result set .. could be large.

**Comment by mjibson** [comment]
Oops! Here we go.

`CREATE VIEW "materialize"."public"."pseudo_source" AS SELECT "column1" AS "key", "column2" AS "value" FROM (VALUES ('a', 1), ('a', 2), ('a', 3), ('a', 4), ('b', 5), ('c', 6), ('c', 7)) AS "tbl"`

**Comment by frankmcsherry** [comment]
Thanks! My quick read is that there is 100% a size of result at which we will fall over, and 7^10 (~ a quarter billion) might be it. Certainly we should do something better, and .. well this will stay open until we do!

**Comment by mjibson** [comment]
Good to know. I'll add this error to my whitelist.

**Comment by ruchirK** [comment]
Hey Matt just want to confirm understanding -- this set of queries causes an error but not a crash right? When I run these queries on most recent master I don't see a crash. 

Also, when we run Smith against Materialize on nightly / ci tests the runner specifically looks for crashes, and ignores errors (given that theres a lot of expected errors in the test queries). Is there something smarter we could be doing with regards to determining which errors are expected vs which are not? We don't currently have a structured format for errors so its hard to determine which are internal vs which are not.

**Comment by mjibson** [comment]
Yes, you have to whitelist errors. Or search for some known internal error string that indicates something wasn't a SQL error. This is listed as step 6 of https://jibson.dev/smith/

> If the returned error is unexpected (perhaps it panic’d, the panic was caught and returned to the user as an internal error), log this query.

**Comment by ruchirK** [comment]
Right, but is the whitelist you are referring to something that is a part of Smith, or is that something we would have to build into the runner ourself? I think that's reasonable to expect users to track which errors are SQL related vs which are not unfortunately that will require some effort to do in materialize at the moment as we just return strings for errors at the moment without much structure.

**Comment by mjibson** [comment]
Yes, this whitelist must be built up by you. I don't recommend running smithers in CI, should be only nightly. Where I've done this before we made it work by doing it the other way around: any non SQL error was flagged as an "internal error", and so we were able to search for that string which indicated a programming error. Both techniques work fine. Do what works best for you.

**Comment by ruchirK** [comment]
@mjibson I've created #3766 to track the next round of changes we need to the Smith runner
